### ISSUE TITLE
[UX] Increase the width of the admin bar search results list (so that less results wrap in two lines).

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1001]
This is why:

![admin_bar-search_results-narrow](https://cloud.githubusercontent.com/assets/2423362/8142971/0509083c-11d8-11e5-855d-c85744229aa9.png)

I was thinking perhaps increase the width to 300px? (this will BTW also accommodate some of the changes I proposed [here](https://github.com/backdrop/backdrop-issues/issues/525#issuecomment-111666887)):

![admin_bar-search_results-wide](https://cloud.githubusercontent.com/assets/2423362/8142972/1f758f42-11d8-11e5-856f-ed3e815b8da7.png)

There should be no side-effect because there's room to the right due to the `.admin-bar-users`/`.admin-bar-account`/`.admin-bar-logout` items next to the search field.

**Comment by klonos** [comment]
I've also played around with setting the width to ```min-width: 160px;``` (same as ```#admin-bar .dropdown li ul```) and the results list got a "flexible" width as I typed.

If you want to give this a spin, then make the following changes in admin_bar.css...

line 137:

```
#admin-bar .dropdown li {
  background-image: none;
- float: left; /* LTR */ 
  height: 100%;
  list-style-image: none;
  list-style-type: none;
  margin: 0 !important;
  padding: 0;
}
```

line 151:
```
#admin-bar .dropdown li li {
- width: 160px;
+ min-width: 160px;
}
```

line 161:

```
#admin-bar .dropdown li ul {
  display: none;
  left: -999em; /* LTR */
  line-height: 1.2em;
  margin: 0;
  position: absolute;
- width: 160px;
+ min-width: 160px
}
```

line 288:

```
#admin-bar .admin-bar-search-results,
#admin-bar .admin-bar-search-results li {
- width: 186px;
+ min-width: 160px;
}
```

**Comment by quicksketch** [comment]
This seems like a good idea to me. The mobile version of the search might need a little attention as well, so if necessary we can use a media query to target larger screens here.

**Comment by klonos** [comment]
This change did have a minor side effect with the "Reports" menu (the expandable icons) - no other menus were effected:

![admin_bar-reports_menu-width_issue](https://cloud.githubusercontent.com/assets/2423362/8143142/c407b99c-11e0-11e5-8a36-6e7a62664e25.png)

Additional changes in the admin_bar.css to fix this:

line 175:
```
#admin-bar .dropdown li li.expandable ul {
- margin: -32px 0 0 160px;
+ margin: -32px 0 0 179px;
}
```

line 193:
```
#admin-bar .dropdown li li.expandable > a {
- background: #45454A url(../images/arrow.png) no-repeat 145px center;
- background: #45454A url(../images/arrow.png) no-repeat 165px center;
}
```

But once you make these changes and the "Reports" menu is fixed, things break for the rest of the 2nd level menus. Weird?:

![admin_bar-config_menu-width_issue](https://cloud.githubusercontent.com/assets/2423362/8143166/0cf5319c-11e2-11e5-8d3c-17bee9949180.png)

Perhaps someone else with a cleared head can sort this out. I hope I helped.

**Comment by klonos** [comment]
Yeah, this improved the situation a lot because search results now do not wrap in a second line (so the results list can hold more entries), but still does not solve the issue of too long menus in admin bar... hence the other issue.

**Comment by klonos** [comment]
Now with #948 implemented, we have even more search results wrapping in two lines.

**Comment by klonos** [comment]
@quicksketch I believe that I am really close to sort this, but I need some JS help.

The current implementation in https://github.com/backdrop/backdrop/blob/1.x/core/modules/admin_bar/js/admin_bar.js#L271 makes it so that `<a>` and `<span>` tags get the class `expanded-trail`. I need that class to be added to the respective `<li>`s as well. I tried playing a bit with it, but being a total JS noob, I can't get it to work. Help please?

**Comment by quicksketch** [comment]
It might be a bug in that JS. I think it's already supposed to add the `expanded-trail` all the way down the menu, not just the lowest level. I think something like this might do it:

Old:
```
    // Add classes for the expanded trail of links.
    $childList
      .parents('ul').addBack().addClass('expanded')
      .siblings('a, span').addClass('expanded-trail');
```

New:
```js
    // Add classes for the expanded trail of links.
    $childList
      .parents('ul').addBack().addClass('expanded').each(function() {
        $(this).siblings('a, span').addClass('expanded-trail');
      });
```

**Comment by klonos** [comment]
> It might be a bug in that JS. I think it's already supposed to add the `expanded-trail` all the way down the menu, not just the lowest level.

Yep, seems like a bug. The code you proposed did not fix it unfortunately, so I filed a separate issue for it. Guess I'll have to delve into JS. I wish the day had 48 hours and that there was an extra day in they week.

**Comment by klonos** [comment]
...after 2 years+ my wish has expanded to "...and another month in the year" 😞

**Comment by BWPanda** [comment]
I made a new PR that fixes this simply and easily: https://github.com/backdrop/backdrop/pull/3254

The only issue is that when the search results drop-down doesn't fit on the screen (e.g. when the 'User counts' and 'Account links' are hidden, or on smaller screens) then the drop-down goes off the screen and is partly unreadable. However, this is a general issue with the admin menu,as the same thing happens with the Reports menu at certain screen sizes, so I think that needs addressing seperately.

**Comment by klonos** [comment]
Well, it doesn't fix it; it improves things. The list still expands off the bottom of the screen, but less 👍 

> the drop-down goes off the screen and is partly unreadable.

Yes, we are trading one problem with another 😅 

![image](https://user-images.githubusercontent.com/2423362/90425545-979dbe80-e0c8-11ea-8ca8-4843607f4412.png)

**Comment by BWPanda** [comment]
> The list still expands off the bottom of the screen, but less

You can have each menu item only take up one line, but still have it go off the bottom of the screen, simply because there are so many results. If that's the problem you're actually trying to fix, I suggest re-titling this issue (I thought it was just about making the menu wider to look better).

As for the horizontal scrollbar, that's already an issue (see other one I filed). So I see this as worth committing, then fixing the other issue seperately (rather than trying to fix all issues here).

**Comment by klonos** [comment]
> I thought it was just about making the menu wider to look better

Yes, that's the goal, and I actually proposed increasing the width in my original report, same as you've done in your PR:

>I was thinking perhaps increase the width to 300px?

...the actual problem is with the inability to scroll off viewport. Perhaps it'd be fixed if we also increased the width of the search box to match the increased width of the results, and then also made sure that the searchbox/results combo stays within viewport _horizontally_?

**Comment by bugfolder** [comment]
Tested: it does what was expected (menu is wider). As noted above, it's still possible for the menu to go off-screen to the right, but that's grist for a future issue. Code reviewed, LGTM.

**Comment by quicksketch** [comment]
Thanks for reviewing! I re-merged 1.x into the PR to see if #4549 solves the problem of search going off-screen.

**Comment by quicksketch** [comment]
Nope, search still going off-screen even after #4549.

Well, although this does not entirely fix the problem and sort of makes mobile worse while improving desktop, I think @klonos and @bugfolder are both agreeing that it's overall an improvement. I'm fine with this being limited scope and we can further improve in separate issues.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/3254 into 1.x and 1.21.x. Let's open new issues as needed for vertical scrolling and keeping search from going off screen on mobile.
